src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py

```diff
@@ -79,66 +70,61 @@ class KalmanODESolution(ODESolution):
-        self.proj_to_dy = kalman_posterior.transition.proj2coord(coord=1)
+        self.proj_to_y = self.kalman_posterior.transition.proj2coord(coord=0)
+        self.proj_to_dy = self.kalman_posterior.transition.proj2coord(coord=1)
 
-    @property
-    def t(self) -> np.ndarray:
-        return self.kalman_posterior.locations
-
-    @cached_property
-    def y(self) -> pnrv_list._RandomVariableList:
-        y_rvs = [
-            _project_rv(self.proj_to_y, rv) for rv in self.kalman_posterior.state_rvs
-        ]
-        return pnrv_list._RandomVariableList(y_rvs)
-
-    @cached_property
-    def dy(self) -> pnrv_list._RandomVariableList:
-        dy_rvs = [
-            _project_rv(self.proj_to_dy, rv) for rv in self.kalman_posterior.state_rvs
-        ]
-        return pnrv_list._RandomVariableList(dy_rvs)
-
-    def __call__(
-        self, t: float
-    ) -> typing.Union[randvars.RandomVariable, pnrv_list._RandomVariableList]:
-        out_rv = self.kalman_posterior(t)
-
-        if np.isscalar(t):
-            return _project_rv(self.proj_to_y, out_rv)
-
-        return pnrv_list._RandomVariableList(
-            [_project_rv(self.proj_to_y, rv) for rv in out_rv]
+        states = _randomvariablelist._RandomVariableList(
+            [_project_rv(self.proj_to_y, rv) for rv in self.kalman_posterior.states]
+        )
+        derivatives = _randomvariablelist._RandomVariableList(
+            [_project_rv(self.proj_to_dy, rv) for rv in self.kalman_posterior.states]
         )
+        super().__init__(
+            locations=kalman_posterior.locations, states=states, derivatives=derivatives
+        )
+
+    def interpolate(self, t: FloatArgType) -> randvars.RandomVariable:
+        out_rv = self.kalman_posterior.interpolate(t)
+        return _project_rv(self.proj_to_y, out_rv)
 
     def sample(
         self,
-        t: typing.Optional[float] = None,
-        size: typing.Optional[probnum.type.ShapeArgType] = (),
+        t: Optional[DenseOutputLocationArgType] = None,
+        size: Optional[ShapeArgType] = (),
+        random_state: Optional[RandomStateArgType] = None,
     ) -> np.ndarray:
-        """Sample from the Gaussian filtering ODE solution by sampling from the Gauss-
-        Markov posterior."""
-        size = probnum.utils.as_shape(size)
 
-        # implement only single samples, rest via recursion
-        # We cannot 'steal' the recursion from self.kalman_posterior.sample,
-        # because we need to project the respective states out of each sample.
-        if size != ():
-            return np.array([self.sample(t=t, size=size[1:]) for _ in range(size[0])])
+        samples = self.kalman_posterior.sample(
+            t=t, size=size, random_state=random_state
+        )
+
+        # Project the samples down to the "true" KalmanODESolution dimensions
+        # (which are a subset of the KalmanPosterior dimensions)
+        ode_samples = np.einsum("dq,...q->...d", self.proj_to_y, samples)
```

Comment:
Could you elaborate what happens here? I know about einsums, but I don't understand why we use it here (or why it even works). Would `[_project_rv(self.proj_to_y, s) for s in samples]` do the same?

Reply:
yes, it would do the same. Einsum is faster though, because it is vectorised :)